// node-odata/lib/resource.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _lodash = require('lodash');

var _model = require('./model');

var _model2 = _interopRequireDefault(_model);

var _rest = require('./rest');

var _rest2 = _interopRequireDefault(_rest);

var Resource = (function () {
  function Resource(name, model) {
    _classCallCheck(this, Resource);

    this._name = name;
    this._url = name;
    this._model = model;
    this._rest = {
      list: {},
      get: {},
      post: {},
      put: {},
      'delete': {} };
    this._actions = {};
    this._options = {};
  }

  _createClass(Resource, [{
    key: 'action',
    value: function action(url, fn, auth) {
      this._actions[url] = fn;
      this._actions[url].auth = auth;
      return this;
    }
  }, {
    key: 'maxTop',
    value: function maxTop(count) {
      this._maxTop = count;
      return this;
    }
  }, {
    key: 'maxSkip',
    value: function maxSkip(count) {
      this._maxSkip = count;
      return this;
    }
  }, {
    key: 'orderBy',
    value: function orderBy(field) {
      this._orderby = field;
      return this;
    }
  }, {
    key: 'list',
    value: function list() {
      this._currentMethod = 'list';
      return this;
    }
  }, {
    key: 'get',
    value: function get() {
      this._currentMethod = 'get';
      return this;
    }
  }, {
    key: 'post',
    value: function post() {
      this._currentMethod = 'post';
      return this;
    }
  }, {
    key: 'put',
    value: function put() {
      this._currentMethod = 'put';
      return this;
    }
  }, {
    key: 'delete',
    value: function _delete() {
      this._currentMethod = 'delete';
      return this;
    }
  }, {
    key: 'all',
    value: function all() {
      //TODO
      throw new Error('Not implemented');
    }
  }, {
    key: 'before',
    value: function before(fn) {
      this._rest[this._currentMethod].before = fn;
      return this;
    }
  }, {
    key: 'after',
    value: function after(fn) {
      this._rest[this._currentMethod].after = fn;
      return this;
    }
  }, {
    key: 'auth',
    value: function auth(fn) {
      this._rest[this._currentMethod].auth = fn;
      return this;
    }
  }, {
    key: 'url',
    value: function url(_url) {
      this._url = _url;
      return this;
    }
  }, {
    key: '_router',
    value: function _router(db) {
      var setting = arguments[1] === undefined ? {} : arguments[1];

      // remove '/' if url is startwith it.
      if (this._url.indexOf('/') === 0) {
        this._url = this._url.substr(1);
      }

      // not allow contain '/' in url.
      if (this._url.indexOf('/') >= 0) {
        throw new Error('Url of resource[' + this._name + '] can\'t contain "/", it can only be allowed to exist in the beginning.');
      }

      _model2['default'].register(db, this._url, this._model);

      var params = {
        options: {
          maxTop: (0, _lodash.min)([setting.maxTop || 100000, this._maxTop || 100000]),
          maxSkip: (0, _lodash.min)([setting.maxSkip || 100000, this._maxSkip || 100000]),
          orderby: this._orderby },
        rest: this._rest,
        actions: this._actions };

      return _rest2['default'].getRouter(db, this._url, params);
    }
  }]);

  return Resource;
})();

exports['default'] = Resource;
module.exports = exports['default'];
